src/renderer/object/instanced_model.rs

```diff
@@ -49,53 +52,173 @@ impl<M: ForwardMaterial> InstancedModel<M> {
+        self.transformations.push(*transformation);
+    }
+
+    ///
+    /// Clear transformations array.
+    ///
+    pub fn clear_transformations(&mut self) {
+        self.transformations = Vec::new();
+        self.update_buffers();
+    }
+
+    ///
+    /// Update instance transform and uv buffers and aabb.
+    ///
+    pub fn update_buffers(&mut self) {
+        self.instance_count = self.transformations.len() as u32;
+        let transformations = self.transformations.as_slice();
         let mut row1 = Vec::new();
         let mut row2 = Vec::new();
         let mut row3 = Vec::new();
-        for transform in transformations {
-            row1.push(transform.x.x);
-            row1.push(transform.y.x);
-            row1.push(transform.z.x);
-            row1.push(transform.w.x);
-
-            row2.push(transform.x.y);
-            row2.push(transform.y.y);
-            row2.push(transform.z.y);
-            row2.push(transform.w.y);
-
-            row3.push(transform.x.z);
-            row3.push(transform.y.z);
-            row3.push(transform.z.z);
-            row3.push(transform.w.z);
+        let mut subt = Vec::new();
+        if self.transformations.len() > 0 {
+            for transform in transformations {
+                row1.push(transform.x.x);
+                row1.push(transform.y.x);
+                row1.push(transform.z.x);
+                row1.push(transform.w.x);
+
+                row2.push(transform.x.y);
+                row2.push(transform.y.y);
+                row2.push(transform.z.y);
+                row2.push(transform.w.y);
+
+                row3.push(transform.x.z);
+                row3.push(transform.y.z);
+                row3.push(transform.z.z);
+                row3.push(transform.w.z);
+
+                for i in [0.0f32, 0.0, 1.0, 1.0] {
+                    subt.push(i);
+                }
+            }
+        } else {
+            row1.push(self.transformation.x.x);
+            row1.push(self.transformation.y.x);
+            row1.push(self.transformation.z.x);
+            row1.push(self.transformation.w.x);
+
+            row2.push(self.transformation.x.y);
+            row2.push(self.transformation.y.y);
+            row2.push(self.transformation.z.y);
+            row2.push(self.transformation.w.y);
+
+            row3.push(self.transformation.x.z);
+            row3.push(self.transformation.y.z);
+            row3.push(self.transformation.z.z);
+            row3.push(self.transformation.w.z);
+        }
+        if self.subtextures.is_some() {
+            for (i, subtex) in self.subtextures.as_ref().unwrap().iter().enumerate() {
+                if i >= transformations.len() {
+                    break;
+                }
+                subt[i * 4] = subtex.x;
+                subt[i * 4 + 1] = subtex.y;
+                subt[i * 4 + 2] = subtex.w;
+                subt[i * 4 + 3] = subtex.h;
+            }
+        } else if self.subtexture.is_some() {
```

Comment:
Same as with the transformations, there should be one global and also one local for each instance.